Feature/custom nonce

rickycodes opened the pull request:
<!--
Thanks for your contribution!

Please ensure that any applicable requirements below are satisfied before submitting this pull request. This will help ensure a quick and efficient review cycle.
-->

**Description**

Opening this as a Draft as it relies on https://github.com/MetaMask/controllers/pull/381

#### This adds a new setting under Advanced that lets you edit the nonce rather than getting it from the network:

![image](https://user-images.githubusercontent.com/675259/110887947-36824600-82b9-11eb-890d-ec84c51d7d7a.png)

#### Displays suggested nonce when you do a new send:

![image](https://user-images.githubusercontent.com/675259/110888020-54e84180-82b9-11eb-8eda-a7aaa402de98.png)

#### which can then be edited:

![image](https://user-images.githubusercontent.com/675259/110888045-63cef400-82b9-11eb-8b5d-c6fd9bcbf0b0.png)

![image](https://user-images.githubusercontent.com/675259/110888080-734e3d00-82b9-11eb-9ca9-0d46ba15f89a.png)


**Checklist**

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

**Issue**

Resolves #2126

rickycodes committed:
Add custom nonce switch

rickycodes committed:
update snapshot

rickycodes committed:
Use props

rickycodes committed:
Get edit nonce button in place

rickycodes committed:
Add edit nonce modal

rickycodes committed:
Add translation strings

rickycodes committed:
Add CustomNonceModal component

rickycodes committed:
Add proposedNonce

rickycodes committed:
undo

rickycodes committed:
Update styles and translation strings

rickycodes committed:
Use a hook

rickycodes committed:
Save nonce to state

rickycodes committed:
Add IncrementDecrementSvg component

rickycodes committed:
Add comment for newly added prop

rickycodes committed:
Fix width @ 80

rickycodes committed:
Add component tests

rickycodes committed:
Display warning

rickycodes committed:
Add tests

rickycodes committed:
Update tests

rickycodes committed:
Add getProposedNonce

rickycodes committed:
Merge some utils and add tests

rickycodes committed:
Add constants

rickycodes committed:
Use object property shorthand

rickycodes committed:
Fix displayWarning to account for String

rickycodes committed:
Add ModalDragger

rickycodes committed:
Add save and cancel

rickycodes committed:
Use the TransactionController from "@metamask/controllers"

rickycodes committed:
Only use custom nonce if showCustomNonce

rickycodes committed:
Update snapshot

rickycodes committed:
Start using Base/Text

rickycodes committed:
Add colors to Base/Text component

rickycodes committed:
Account for NaN values

rickycodes committed:
Use Base/Text

wachunei commented on app/components/Views/SendFlow/Confirm/index.js:
```diff
@@ -1029,6 +1072,20 @@ class Confirm extends PureComponent {
+								<Text bold black>
+									{strings('transaction.custom_nonce')}
+								</Text>
+								<Text bold blue>
```
Thanks for adding blue 👌, however to keep things semantically correct I'd use `bold link` here, what do you think?

rickycodes replied:
oh jeez! I didn't see that. thanks for the heads up. I'll change that.

rickycodes committed:
Move increment decrement buttons down

rickycodes committed:
Update tests

rickycodes committed:
Add showSoftInputOnFocus={false}

rickycodes committed:
Update tests

rickycodes committed:
Add icon

rickycodes committed:
Use link

rickycodes committed:
Update tests

rickycodes committed:
prep for QA

cjeria commented:
For the warning copy, instead of repeating what is already mentioned in the default copy, I'm now thinking we should mention what _could_ happen if they continue with the higher than suggested nonce. 

How about something like this instead? cc @rickycodes @omnat @jakehaugen 
![image](https://user-images.githubusercontent.com/597770/111521252-2c77a180-8716-11eb-8632-1b2a0c5aa48e.png)

rickycodes committed:
Update warning text

Gudahtt commented on package.json:
```diff
@@ -74,7 +74,7 @@
     "@exodus/react-native-payments": "https://github.com/wachunei/react-native-payments.git#package-json-hack",
     "@metamask/contract-metadata": "^1.23.0",
-    "@metamask/controllers": "^6.1.1",
+    "@metamask/controllers": "file:./metamask-controllers-v6.1.2.tgz",
```
This feature has been published in v9.2.0 of the controllers repo, so this can be updated now to reference the published version

rickycodes committed:
Use 6.2.0 controllers

rickycodes committed:
Update translation strings

rickycodes committed:
update snapshot

rickycodes committed:
Remove TODO

rickycodes committed:
Use EvilIcons

rickycodes committed:
Update snapshot

rickycodes committed:
Beef up tests for toLowerCaseCompare

estebanmino commented on app/core/Engine.js:
```diff
@@ -18,7 +18,7 @@ import {
 } from '@metamask/controllers';
 
-import { SwapsController } from '@estebanmino/controllers';
+import { /* TransactionController ,*/ SwapsController } from '@estebanmino/controllers';
```
this is not needed anymore `/* TransactionController ,*/`

estebanmino commented on app/util/transactions.js:
```diff
@@ -424,5 +425,26 @@ export function getNormalizedTxState(state) {
+
+	return Math.max(...nonces) + 1;
+};
+
```
rather than taking this from the transactions list I'd try to query the blockchain instead with [getTransactionCount](https://infura.io/docs/ethereum/json-rpc/eth-getTransactionCount). We can miss transactions that didn't happen from the app, specially for custom networks added by users.

We can do it directly with `query(this.ethQuery, 'getTransactionCount', [from, 'pending']))` from the app.

WDYT?

rickycodes replied:
personally, I like the simplicity of this approach (one less blockchain query)... but maybe it makes sense to _also_ get the nonce from the network and then use that _if_ it's different from the local one?

estebanmino replied:
I can imagine txs in custom networks that aren't on the app, if they want to use this feature it won't work for them

estebanmino replied:
this is also important for users with 3rd party apis turned off, it will make this feature fail if txs happened somewhere else

estebanmino replied:
WDYT?

Gudahtt replied:
We rely upon just the transaction list in the extension, but we're in a different situation. We get our list of confirmed transactions directly from the network (not Etherscan), and we update it every single block (whereas mobile seems to get an updated transaction list according to certain UI actions - not sure which).

Barring some major change to how the `@metamask/controllers` transaction controller works, we'll need to use `getTransactionCount` here I think.

rickycodes committed:
Remove comment

rickycodes committed:
Merge branch 'develop' into feature/custom-nonce

rickycodes committed:
Merge branch 'develop' into feature/custom-nonce

rickycodes committed:
update snapshot

rickycodes committed:
Get nonce from network

rickycodes committed:
Merge branch 'develop' into feature/custom-nonce

rickycodes committed:
Update snapshot

ibrahimtaveras00 reviewed (changes requested):
**Issue 1:**

So far this is only implemented on send flow, but also needs to be implemented on dapp transactions which is currently not the case

refer to = https://consensys.slack.com/archives/GBW7S9FSN/p1617653913259200?thread_ts=1617652943.258500&cid=GBW7S9FSN

**Issue 2:**

Some text changes compared to mock up

![image](https://user-images.githubusercontent.com/41752922/113626013-9dccb500-962f-11eb-992e-65c301c3ab40.png)

**Issue 3:**

When you submit another txn with the same nonce of previous txn to cancel it out, it still shows up in txn history as submitted even though it was dropped and replaced

![image](https://user-images.githubusercontent.com/41752922/113629240-24839100-9634-11eb-973c-1d69f9ecbe95.png)

steps:
- submit a txn with low gas (take note of nonce)
- submit a new txn with higher gas, but use the same exact nonce of previous txn
- once the new txn gets confirmed the older txn still shows as submitted (even after killing app and relaunching, or reloading if you're on debug, it still shows as submitted)

rickycodes committed:
Merge branch 'develop' into feature/custom-nonce

rickycodes committed:
fix lint nits

rickycodes committed:
fix copy

rickycodes committed:
Update snapshot

rickycodes committed:
Move edit nonce functionality into TransactionReviewFeeCard

rickycodes committed:
Add custom nonce to dapp send

rickycodes committed:
Use redux actions

rickycodes committed:
Continue using actions

rickycodes committed:
Set custom nonce on confirm

rickycodes committed:
Update snapshot

rickycodes committed:
Move nonce and proposedNonce off transaction object

rickycodes committed:
Bump controllers

rickycodes committed:
Merge branch 'develop' into feature/custom-nonce

rickycodes committed:
Only getNetworkNonce if proposedNonce is not set

rickycodes committed:
fix props

rickycodes committed:
delete tgz

rickycodes committed:
Add getNetworkNonce to networks utility

rickycodes committed:
Merge branch 'develop' into feature/custom-nonce

* develop:
  v2.1.0 (#2481)
  Analytics v2 (priority 1) (#2456)
  Fix/gas estimations (#2408)

rickycodes committed:
document props

rickycodes committed:
Update snapshots

rickycodes commented:
I've gone ahead and removed the design qa label since I've gone over this with Christian and he's okay'd it all. re: https://github.com/MetaMask/metamask-mobile/pull/2371#pullrequestreview-628156144 I've addressed all QA feedback except for issue 3 as discussed. this is ready for qa again.

estebanmino commented on app/components/UI/CustomNonceModal/index.js:
```diff
@@ -0,0 +1,224 @@
+		onChangeText(newValue > 1 ? newValue : 1);
+	};
+
+	const saveAndClose = value => {
```
just personal preference, maybe it would be cleaner to use the nonce from state here directly rather than calling the method with the value from render https://github.com/MetaMask/metamask-mobile/pull/2371/files#diff-7358361397d4cbafe60eca2a1bae2ed5226b676ef85d91995e272bdac2159682R205

estebanmino commented on app/components/UI/TransactionReview/TransactionReviewInformation/index.js:
```diff
@@ -184,13 +186,52 @@ class TransactionReviewInformation extends PureComponent {
+			<CustomNonceModal
+				proposedNonce={proposedNonce}
+				nonceValue={nonce}
+				close={() => this.toggleNonceModal()}
```
```suggestion
				close={this.toggleNonceModal}
```

estebanmino commented on app/util/networks.js:
```diff
@@ -165,3 +167,9 @@ export function isPrefixedFormattedHexString(value) {
+export const getNetworkNonce = async ({ from }) => {
+	const { TransactionController } = Engine.context;
+	const networkNonce = await util.query(TransactionController.ethQuery, 'getTransactionCount', [from, 'pending']);
+	return parseInt(networkNonce, 16);
```
nice and clean 👌

rickycodes committed:
Update app/components/UI/TransactionReview/TransactionReviewInformation/index.js

Co-authored-by: Esteban Miño <efmino@uc.cl>

rickycodes committed:
use nonce directly instead of passing it in render

rickycodes committed:
Merge branch 'feature/custom-nonce' of github.com:MetaMask/metamask-mobile into feature/custom-nonce

* 'feature/custom-nonce' of github.com:MetaMask/metamask-mobile:
  Update app/components/UI/TransactionReview/TransactionReviewInformation/index.js

estebanmino reviewed (approved):
lgtm!

rickycodes committed:
Merge branch 'develop' into feature/custom-nonce

rickycodes committed:
Merge branch 'develop' into feature/custom-nonce

ibrahimtaveras00 reviewed (approved):
All issues resolved, QA Passed 👍🏽

rickycodes committed:
Merge branch 'develop' into feature/custom-nonce
